### PR TITLE
#166107096 Persist user controller 

### PR DESCRIPTION
**Pull request by AnselemOdims**
#### What does this PR do?
Persists the User controller with the database
#### Description of Task to be completed?
- Modify and add new tests
- Modify the user controller for the register user endpoint
- Modify the controller for the login endpoint 
- Modify the controller for the verify a user endpoint
- Change the controller for the get a specific user endpoint
- Change the controller for the get all users endpoint
- Introduce a delete user controller
- Modify validations to query database 
- Seed Admin and Users in the database
- Adjust the test script in the package.json
- Modify the travis.yml file so travis can build properly
- Introduce new helper functions
- Ensure that the under-listed endpoints are working correctly
```
POST /api/v1/auth/signup
POST /api/v1/auth/login
PATCH /api/v1/users/:email/verify
GET /api/v1/users
GET /api/v1/users/:id
DELETE /api/v1/users/:id
```
#### How should this be manually tested?
- Clone the repository
- cd into loaner
- Run git checkout ft-persist-user-controller-166107096 
- Run npm install in the command line
- create a .env file and set the following
```
  SECRET_KEY = secret key to create a jwt token
  ADMIN_PASSWORD = seeded admin password
  USER_PASSWORD = seeded user password
  DATABASE_URL = the database url
  TESTDB_URL = the test database url
```
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[Delivers #166107096](https://www.pivotaltracker.com/n/projects/2328795)

### CODE REVIEW

File: server/controllers/userController.js [hunk 1]
```diff
@@ -18,18 +24,30 @@ class UserController {
...
     });
+    return res
+      .status(201)
+      .json({
+        status: 201,
+        message: 'User Registration Successful!',
+        success: true,
+        data: {
+          token,
+          ...user,
```
Comment (V-84) [inline server/controllers/userController.js:48]: remove this user information and store it in the token

File: server/controllers/userController.js [hunk 2]
```diff
@@ -40,18 +58,32 @@ class UserController {
...
+      .status(200)
+      .json({
+        status: 200,
+        message: 'Log in Successful!',
+        success: true,
+        data: {
+          token,
+          id: user.id,
+          firstName: user.firstname,
+          lastName: user.lastname,
```
Comment (V-84) [inline server/controllers/userController.js:80]: please remove this user info and put them in the token

File: server/controllers/userController.js [hunk 3]
```diff
@@ -63,13 +95,24 @@ class UserController {
    */
   static async verifyUser(req, res) {
     const { email } = req.params;
```
Comment (V-84) [inline server/controllers/userController.js:97]: can you explain what this does

File: server/controllers/userController.js [hunk 4]
```diff
@@ -18,18 +24,30 @@ class UserController {
...
-      status: 201,
-      message: 'User Registration Successful!',
-      data: {
-        token: userToken,
-        user,
-      },
+    const {
+      firstName, lastName, email, address,
+    } = req.body;
+    const password = await Helpers.hashPassword(req.body.password);
```
Comment (V-84) [inline server/controllers/userController.js:30]: put this in a function of is own

File: server/controllers/userController.js [hunk 5]
```diff
@@ -63,13 +95,24 @@ class UserController {
...
   static async verifyUser(req, res) {
     const { email } = req.params;
-    const { status } = req.body;
-    const verifiedUser = await User.verify(email, status, users);
-    return res.status(200).json({
-      status: 200,
-      message: 'User Verification Successful!',
-      data: verifiedUser,
-    });
+    const result = await db.query(verifyUser, [email]);
```
Comment (V-84) [inline server/controllers/userController.js:98]: why are you making 2 database calls.?

File: server/middlewares/repaymentValidation.js [hunk 6]
```diff
@@ -1,85 +1,85 @@
-import Repayments from '../models/repayment';
-import Loans from '../models/loan';
-import Helpers from '../utils/helpers';
-import db from '../models/db';
+// import Repayments from '../models/repayment';
```
Comment (V-84) [inline server/middlewares/repaymentValidation.js:1]: this is bad practice... You were being lazy... Please note this...Learning to make proper use of git would prevent this

File: server/middlewares/userValidation.js [hunk 7]
```diff
@@ -123,23 +127,50 @@ class UserValidation {
...
-    const arr = ['verified', 'unverified'];
-    const user = await Helpers.findByMail(email, users);
+    const arr = ['verified'];
+    const user = await db.query(getUserByEmail, [email]);
+
+    // check if the request body is correct
+    if (!arr.includes(status)) {
+      return res
+        .status(400)
+        .json({ status: 400, error: 'The status can either be verified or unverified' });
```
Comment (V-84) [inline server/middlewares/userValidation.js:139]: this message is confusing `'The status can either be verified or unverified'`

**Review by V-84 — COMMENTED** [review]
please attend to the comments

File: server/controllers/userController.js (hunk 1 above)
Comment (AnselemOdims) [reply server/controllers/userController.js:48]: So I should return just the token in the response object?

File: server/controllers/userController.js (hunk 3 above)
Comment (AnselemOdims) [reply server/controllers/userController.js:97]: This is the route handler for verifying users, after confirming their address

File: server/controllers/userController.js (hunk 4 above)
Comment (AnselemOdims) [reply server/controllers/userController.js:30]: ok

File: server/controllers/userController.js (hunk 5 above)
Comment (AnselemOdims) [reply server/controllers/userController.js:98]: The first call updates the user's status, while the other one gets the updated user

File: server/middlewares/repaymentValidation.js (hunk 6 above)
Comment (AnselemOdims) [reply server/middlewares/repaymentValidation.js:1]: I apologise if I acted lazy and didnt didnt make proper use of git. I will always do better from now on.

File: server/middlewares/userValidation.js (hunk 7 above)
Comment (AnselemOdims) [reply server/middlewares/userValidation.js:139]: Its an error message to let the admin know that a user can only be verified or unverified, if he tries to pass another string to the request body.

**Pull request merged**